### ISSUE TITLE
Symbol::delete_package bug

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4153]
Migrated from [rt.perl.org#7206](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7206) (status was 'rejected')

Searchable as RT7206$

**Comment by p5pRT** [comment]
### From gt4556a@prism.gatech.edu



Problem&#8203;: Symbol\.pm 's delete\_package function doesn't delete the package from
%INC\, so when require\(\)ing the package again\, it's still found in %INC and not
reloaded\.

Fix&#8203;: making delete\_package delete the package from %INC as well or documenting
this so the coder is aware of it and can manually remove the package from %INC\.

Sample code and a patch on request\.

<details><summary>Perl Info</summary>

```


Site configuration information for perl 5.00503:

Configured by willday at Fri Apr  2 11:27:26 EST 1999.

Summary of my perl5 (5.0 patchlevel 5 subversion 3) configuration:
  Platform:
    osname=solaris, osvers=2.6, archname=sun4-solaris
    uname='sunos rom 5.6 generic_105181-12 sun4u sparc sunw,ultra-1 '
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef useperlio=undef d_sfio=undef
  Compiler:
    cc='cc', optimize='-fast', gccversion=
    cppflags=''
    ccflags =''
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-fast -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -lgdbm -ldb -ldl -lm -lc -lcrypt
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-KPIC', lddlflags='-G -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.00503:
    /fuse4/47/gt4556a/perl/sun4-solaris
    /fuse4/47/gt4556a/perl
    /usr/local/lib/perl5/5.00503/sun4-solaris
    /usr/local/lib/perl5/5.00503
    /usr/local/lib/perl5/site_perl/5.005/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.005
    .


Environment for perl 5.00503:
    HOME=/fuse4/47/gt4556a
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/fuse4/47/gt4556a/local/bin:/usr/bin:/usr/ccs/bin:/opt/SUNWspro/bin:/usr/local/bin:.:/tnt1/38/gtwreck/bin:/fuse4/47/gt4556a/bin:/fuse4/47/gt4556a/local/bin:/fuse4/47/gt4556a/nmh/bin:/fuse4/47/gt4556a/nmh/lib
    PERL5LIB=/fuse4/47/gt4556a/perl
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @jhi

Consider this a request\.

> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
> \-\-\-
> Site configuration information for perl 5\.00503&#8203;:
> 
> Configured by willday at Fri Apr  2 11&#8203;:27&#8203;:26 EST 1999\.
> 
> Summary of my perl5 \(5\.0 patchlevel 5 subversion 3\) configuration&#8203;:
>   Platform&#8203;:
>     osname=solaris\, osvers=2\.6\, archname=sun4\-solaris
>     uname='sunos rom 5\.6 generic\_105181\-12 sun4u sparc sunw\,ultra\-1 '
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     usethreads=undef useperlio=undef d\_sfio=undef
>   Compiler&#8203;:
>     cc='cc'\, optimize='\-fast'\, gccversion=
>     cppflags=''
>     ccflags =''
>     stdchar='unsigned char'\, d\_stdstdio=define\, usevfork=false
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>     alignbytes=8\, usemymalloc=y\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='cc'\, ldflags ='\-fast \-L/usr/local/lib'
>     libpth=/usr/local/lib /lib /usr/lib /usr/ccs/lib
>     libs=\-lsocket \-lnsl \-lgdbm \-ldb \-ldl \-lm \-lc \-lcrypt
>     libc=/lib/libc\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
>     cccdlflags='\-KPIC'\, lddlflags='\-G \-L/usr/local/lib'
> 
> Locally applied patches&#8203;:
>     
> 
> \-\-\-
> @&#8203;INC for perl 5\.00503&#8203;:
>     /fuse4/47/gt4556a/perl/sun4\-solaris
>     /fuse4/47/gt4556a/perl
>     /usr/local/lib/perl5/5\.00503/sun4\-solaris
>     /usr/local/lib/perl5/5\.00503
>     /usr/local/lib/perl5/site\_perl/5\.005/sun4\-solaris
>     /usr/local/lib/perl5/site\_perl/5\.005
>     \.
> 
> \-\-\-
> Environment for perl 5\.00503&#8203;:
>     HOME=/fuse4/47/gt4556a
>     LANG \(unset\)
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=/fuse4/47/gt4556a/local/bin&#8203;:/usr/bin&#8203;:/usr/ccs/bin&#8203;:/opt/SUNWspro/bin&#8203;:/usr/local/bin&#8203;:\.&#8203;:/tnt1/38/gtwreck/bin&#8203;:/fuse4/47/gt4556a/bin&#8203;:/fuse4/47/gt4556a/local/bin&#8203;:/fuse4/47/gt4556a/nmh/bin&#8203;:/fuse4/47/gt4556a/nmh/lib
>     PERL5LIB=/fuse4/47/gt4556a/perl
>     PERL\_BADLANG \(unset\)
>     SHELL=/usr/local/bin/zsh

**Comment by p5pRT** [comment]
### From @pudge

> On Fri\, Jun 29\, 2001 at 04&#8203;:30&#8203;:14PM \-0400\, Jörg Ziefle wrote&#8203;:
> > This is a bug report for perl from gt4556a@&#8203;gatech\.edu\,
> > generated with the help of perlbug 1\.26 running under perl 5\.00503\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > Problem&#8203;: Symbol\.pm 's delete\_package function doesn't delete the package 
> > from
> > %INC\, so when require\(\)ing the package again\, it's still found in %INC and 
> > not
> > reloaded\.
> > 
> > Fix&#8203;: making delete\_package delete the package from %INC as well or 
> > documenting
> > this so the coder is aware of it and can manually remove the package from 
> > %INC\.
> > 
> > Sample code and a patch on request\.
> 
> Consider this a request\.

Personally\, I would prefer a doc patch\.  When I've used delete\_package 
before\, I've not wanted anything deleted from %INC\.  At most\, make the 
behavior optional\.  It'd be hard to write code for it anyway; what if I 
want to delete the package Tie&#8203;::RefHash&#8203;::Nestable?  There's no 
accompanying file for that\.  Or\, conversely\, if I delete the package 
\(and file from %INC\) for Tie&#8203;::RefHash\, the package 
Tie&#8203;::RefHash&#8203;::Nestable will still be around\.

delete\_package should not be used to delete a module or library\, and I 
don't think we should conflate the two\.  It is there to just clear out a 
namespace\, which could be used in conjunction with delete'ing from %INC\.  
Perhaps there's room for someone to write an "unuse" pragma?  It'd be 
pretty hairy\, I think\, to completely back out of a "use"\.  How does Perl 
know which symbols have been exported to which packages?  Which packages 
are in which files?  It seems like someone wants to do with 
delete\_package more than it is designed to do\, and even doing all that 
is desired would be difficult anyway\.

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Jul 31 22&#8203;:28&#8203;:21 2001\, RT\_System wrote&#8203;:
> > On Fri\, Jun 29\, 2001 at 04&#8203;:30&#8203;:14PM \-0400\, J�rg Ziefle wrote&#8203;:
> > > This is a bug report for perl from gt4556a@&#8203;gatech\.edu\,
> > > generated with the help of perlbug 1\.26 running under perl
> 5\.00503\.
> > >
> > >
> > > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > > \[Please enter your report here\]
> > >
> > > Problem&#8203;: Symbol\.pm 's delete\_package function doesn't delete the
> package
> > > from
> > > %INC\, so when require\(\)ing the package again\, it's still found in
> %INC and
> > > not
> > > reloaded\.
> > >
> > > Fix&#8203;: making delete\_package delete the package from %INC as well or
> > > documenting
> > > this so the coder is aware of it and can manually remove the
> package from
> > > %INC\.
> > >
> > > Sample code and a patch on request\.
> >
> > Consider this a request\.
> 
> Personally\, I would prefer a doc patch\.  When I've used delete\_package
> before\, I've not wanted anything deleted from %INC\.  At most\, make the
> behavior optional\.  It'd be hard to write code for it anyway; what if
> I
> want to delete the package Tie&#8203;::RefHash&#8203;::Nestable?  There's no
> accompanying file for that\.  Or\, conversely\, if I delete the package
> \(and file from %INC\) for Tie&#8203;::RefHash\, the package
> Tie&#8203;::RefHash&#8203;::Nestable will still be around\.
> 
> delete\_package should not be used to delete a module or library\, and I
> don't think we should conflate the two\.  It is there to just clear out
> a
> namespace\, which could be used in conjunction with delete'ing from
> %INC\.
> Perhaps there's room for someone to write an "unuse" pragma?  It'd be
> pretty hairy\, I think\, to completely back out of a "use"\.  How does
> Perl
> know which symbols have been exported to which packages?  Which
> packages
> are in which files?  It seems like someone wants to do with
> delete\_package more than it is designed to do\, and even doing all that
> is desired would be difficult anyway\.

pudge responded to the original poster's request more than eleven years
ago\, and the original poster has not further responded\.  I believe
pudge's reasoning is sound and recommend that this ticket be closed\.

I am taking the ticket for that purpose and will close it in seven days
unless someone wants to take the discussion over\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Feb 24 15&#8203;:39&#8203;:18 2013\, jkeenan wrote&#8203;:

> 
> pudge responded to the original poster's request more than eleven years
> ago\, and the original poster has not further responded\.  I believe
> pudge's reasoning is sound and recommend that this ticket be closed\.
> 
> I am taking the ticket for that purpose and will close it in seven days
> unless someone wants to take the discussion over\.
> 
> Thank you very much\.
> Jim Keenan

Closing as per schedule\.

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @rurban

On Feb 24\, 2013\, at 5&#8203;:39 PM\, James E Keenan via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Tue Jul 31 22&#8203;:28&#8203;:21 2001\, RT\_System wrote&#8203;:
>>> On Fri\, Jun 29\, 2001 at 04&#8203;:30&#8203;:14PM \-0400\, J�rg Ziefle wrote&#8203;:
>>>> This is a bug report for perl from gt4556a@&#8203;gatech\.edu\,
>>>> generated with the help of perlbug 1\.26 running under perl
>> 5\.00503\.
>>>> 
>>>> 
>>>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>>>> \[Please enter your report here\]
>>>> 
>>>> Problem&#8203;: Symbol\.pm 's delete\_package function doesn't delete the
>> package
>>>> from
>>>> %INC\, so when require\(\)ing the package again\, it's still found in
>> %INC and
>>>> not
>>>> reloaded\.
>>>> 
>>>> Fix&#8203;: making delete\_package delete the package from %INC as well or
>>>> documenting
>>>> this so the coder is aware of it and can manually remove the
>> package from
>>>> %INC\.
>>>> 
>>>> Sample code and a patch on request\.
>>> 
>>> Consider this a request\.
>> 
>> Personally\, I would prefer a doc patch\.  When I've used delete\_package
>> before\, I've not wanted anything deleted from %INC\.  At most\, make the
>> behavior optional\.  It'd be hard to write code for it anyway; what if
>> I
>> want to delete the package Tie&#8203;::RefHash&#8203;::Nestable?  There's no
>> accompanying file for that\.  Or\, conversely\, if I delete the package
>> \(and file from %INC\) for Tie&#8203;::RefHash\, the package
>> Tie&#8203;::RefHash&#8203;::Nestable will still be around\.
>> 
>> delete\_package should not be used to delete a module or library\, and I
>> don't think we should conflate the two\.  It is there to just clear out
>> a
>> namespace\, which could be used in conjunction with delete'ing from
>> %INC\.
>> Perhaps there's room for someone to write an "unuse" pragma?  It'd be
>> pretty hairy\, I think\, to completely back out of a "use"\.  How does
>> Perl
>> know which symbols have been exported to which packages?  Which
>> packages
>> are in which files?  It seems like someone wants to do with
>> delete\_package more than it is designed to do\, and even doing all that
>> is desired would be difficult anyway\.
> 
> pudge responded to the original poster's request more than eleven years
> ago\, and the original poster has not further responded\.  I believe
> pudge's reasoning is sound and recommend that this ticket be closed\.
> 
> I am taking the ticket for that purpose and will close it in seven days
> unless someone wants to take the discussion over\.

The doc patch is still missing\.
See the attached patch

**Comment by p5pRT** [comment]
### From @rurban

<details><summary>0001-7206-document-Symbol-delete_package-does-not-touch-I.patch</summary>

```diff
From 18bc6bf0eb6d7ee9dd3b57fefa7c0e7b2e0f8b79 Mon Sep 17 00:00:00 2001
From: Reini Urban <rurban@x-ray.at>
Date: Thu, 7 Mar 2013 06:54:50 -0600
Subject: [PATCH] #7206 document Symbol::delete_package does not touch %INC

---
 lib/Symbol.pm | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/lib/Symbol.pm b/lib/Symbol.pm
index 1e408b5..ef347d3 100644
--- a/lib/Symbol.pm
+++ b/lib/Symbol.pm
@@ -65,7 +65,9 @@ even if C<use strict 'refs'> is in effect.
 
 C<Symbol::delete_package> wipes out a whole package namespace.  Note
 this routine is not exported by default--you may want to import it
-explicitly.
+explicitly. Also note that this routine does not remove the package
+module name from the C<%INC> hash. So redefinition by reloading via
+C<require> will fail.
 
 =head1 BUGS
 
-- 
1.8.1.3


```
</details>
